lib/Sema/TypeCheckType.cpp

```diff
@@ -2458,6 +2465,11 @@ Type TypeResolver::resolveDictionaryType(DictionaryTypeRepr *repr,
 
 Type TypeResolver::resolveOptionalType(OptionalTypeRepr *repr,
                                        TypeResolutionOptions options) {
+  options |= TR_OptionalTypeArgument;
```

Comment:
Yuck, that array case will fire :-(

I'll have to have the flag be an immediately-in-optional-type-argument, like how we handle tuples as function parameters